src/xdp/program.c

```diff
@@ -1360,51 +1393,30 @@ XdpProgramDetachRxQueue(
     _In_ XDP_PROGRAM_OBJECT *ProgramObject
     )
 {
-    XDP_RX_QUEUE *RxQueue = ProgramObject->RxQueue;
+    XDP_PROGRAM_BINDING *ProgramBinding = ProgramObject->Binding;
+    XDP_RX_QUEUE *RxQueue = ProgramBinding->RxQueue;
+    XDP_PROGRAM *CompiledProgram = NULL;
 
-    TraceEnter(TRACE_CORE, "Program=%p", ProgramObject);
+    TraceEnter(TRACE_CORE, "Detach ProgramObject=%p from RxQueue", ProgramObject);
 
     ASSERT(RxQueue != NULL);
-    ASSERT(!ProgramObject->Flags.IsMetaProgram);
-
-    if (XdpRxQueueGetProgram(RxQueue) == &ProgramObject->Program) {
-        ASSERT(!ProgramObject->Flags.SharingEnabled);
-        XdpRxQueueDeregisterNotifications(RxQueue, &ProgramObject->RxQueueNotificationEntry);
-        XdpRxQueueSetProgram(RxQueue, NULL, NULL, NULL);
-    } else if (ProgramObject->Flags.SharingEnabled && !IsListEmpty(&ProgramObject->SharingLink)) {
-        XDP_PROGRAM *MetaProgram = XdpRxQueueGetProgram(RxQueue);
-        XDP_PROGRAM_OBJECT *MetaProgramObject =
-            CONTAINING_RECORD(MetaProgram, XDP_PROGRAM_OBJECT, Program);
 
+    if (!IsListEmpty(&ProgramBinding->RxQueueEntry)) {
         //
-        // This program isn't directly attached to the RX queue, but a meta
-        // program is. Remove this program from the metaprogram, and either
-        // update the program ruleset (in-place) or remove the meta-program if
-        // this was the final shared program.
+        // Remove the binding from the RX queue and recompile bound programs.
         //
-        ASSERT(MetaProgram != NULL);
-        ASSERT(MetaProgramObject->Flags.IsMetaProgram);
-
-        XdpRxQueueDeregisterNotifications(RxQueue, &ProgramObject->RxQueueNotificationEntry);
-        RemoveEntryList(&ProgramObject->SharingLink);
-        InitializeListHead(&ProgramObject->SharingLink);
+        RemoveEntryList(&ProgramBinding->RxQueueEntry);
+        InitializeListHead(&ProgramBinding->RxQueueEntry);
 
-        if (IsListEmpty(&MetaProgramObject->SharingLink)) {
-            XdpRxQueueSetProgram(RxQueue, NULL, NULL, NULL);
-            ExFreePoolWithTag(MetaProgramObject, XDP_POOLTAG_PROGRAM);
-            TraceInfo(
-                TRACE_CORE, "Detached metaprogram RxQueue=%p Program=%p",
-                RxQueue, MetaProgramObject);
-        } else {
-            XdpRxQueueSync(RxQueue, XdpProgramPopulateMetaProgram, MetaProgramObject);
-            TraceInfo(
-                TRACE_CORE, "Updated metaprogram RxQueue=%p Program=%p",
-                RxQueue, MetaProgramObject);
-        }
+        XdpProgramCompileNewProgram(RxQueue, &CompiledProgram);
```

Comment:
Ignoring return status? Also, now if one program gets removed (e.g. process A exits) then if a memory allocation fails, all other programs on that system are removed, right? That's definitely undesirable.

Reply:
Ah, I have overlooked this case.

Reply:
Fixed it by updating the "compiled" program in-place.